In [8]:
pip install PyMySQL

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [9]:
pip install SQLAlchemy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [10]:
pip install python_dotenv

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [12]:
import dotenv
import os
from sqlalchemy import create_engine
import pandas as pd

username = os.getenv("username")
password = os.getenv("password")
host = os.getenv("host")
dbname = os.getenv("dbname")

dotenv.load_dotenv(override=True)
conn_string = "mysql+pymysql://" + username + ":" + password + "@" + host + "/" + dbname
db_engine = create_engine(conn_string)

query="SELECT * FROM dimproduct"
df=pd.read_sql(query, db_engine)

# Esercizi W11D2

### Quanti dati ci sono in totale?

In [13]:
df.shape

(606, 36)

### Quali sono i metadati?

In [21]:
df.columns

Index(['ProductKey', 'ProductAlternateKey', 'ProductSubcategoryKey',
       'WeightUnitMeasureCode', 'SizeUnitMeasureCode', 'EnglishProductName',
       'SpanishProductName', 'FrenchProductName', 'StandardCost',
       'FinishedGoodsFlag', 'Color', 'SafetyStockLevel', 'ReorderPoint',
       'ListPrice', 'Size', 'SizeRange', 'Weight', 'DaysToManufacture',
       'ProductLine', 'DealerPrice', 'Class', 'Style', 'ModelName',
       'LargePhoto', 'EnglishDescription', 'FrenchDescription',
       'ChineseDescription', 'ArabicDescription', 'HebrewDescription',
       'ThaiDescription', 'GermanDescription', 'JapaneseDescription',
       'TurkishDescription', 'StartDate', 'EndDate', 'Status'],
      dtype='object')

In [23]:
df.index

RangeIndex(start=0, stop=606, step=1)

### Stampiamo il primo elemento

In [26]:
df.head(1)

ProductKey ProductAlternateKey  ProductSubcategoryKey  \
0           1             AR-5381                    NaN   

  WeightUnitMeasureCode SizeUnitMeasureCode EnglishProductName  \
0                  None                None    Adjustable Race   

  SpanishProductName FrenchProductName  StandardCost  FinishedGoodsFlag  ...  \
0                                                NaN                  0  ...   

  ChineseDescription  ArabicDescription  HebrewDescription  ThaiDescription  \
0               None               None               None             None   

  GermanDescription JapaneseDescription  TurkishDescription  StartDate  \
0              None                None                None 2003-07-01   

  EndDate   Status  
0     NaT  Current  

[1 rows x 36 columns]

### Stampiamo l'ultimo elemento

In [29]:
df.tail(1)

ProductKey ProductAlternateKey  ProductSubcategoryKey  \
605         606          BK-R19B-52                    2.0   

    WeightUnitMeasureCode SizeUnitMeasureCode  EnglishProductName  \
605                    LB                  CM  Road-750 Black, 52   

            SpanishProductName           FrenchProductName  StandardCost  \
605  Carretera: 750, negra, 52  Vélo de route 750 noir, 52        343.65   

     FinishedGoodsFlag  ...                 ChineseDescription  \
605                  1  ...  入门级成人自行车；确保越野旅行或公路骑乘的舒适。快拆式车毂和轮缘。   

                                     ArabicDescription  \
605  إنها دراجة مناسبة للمبتدئين من البالغين؛ فهي ت...   

                                     HebrewDescription  \
605  אופני מבוגרים למתחילים; מציעים רכיבה נוחה "מחו...   

                                       ThaiDescription  \
605  จักรยานระดับเริ่มต้นสำหรับผู้ใหญ่ ให้ความสบายใ...   

                                     GermanDescription  \
605  Ein Erwachsenenrad für Einsteiger; bietet Komf...   

                                   JapaneseDescription  \
605  エントリー レベルに対応する、クロスカントリーにも街への買い物にも快適な、大人の自転車。ハブ...   

                                    TurkishDescription  StartDate EndDate  \
605  "Başlangıç seviyesinde yetişkin bisikleti, kır... 2013-07-01     NaT   

      Status  
605  Current  

[1 rows x 36 columns]

### Riusciamo a stampare cinque elementi a caso?

In [33]:
df.sample(5)

ProductKey ProductAlternateKey  ProductSubcategoryKey  \
586         587          BK-M38S-38                    1.0   
296         297          FR-M94B-42                   12.0   
162         163             MS-1981                    NaN   
347         348          BK-M82B-38                    1.0   
503         504          FR-T67U-58                   16.0   

    WeightUnitMeasureCode SizeUnitMeasureCode             EnglishProductName  \
586                    LB                  CM      Mountain-400-W Silver, 38   
296                    LB                  CM  HL Mountain Frame - Black, 42   
162                  None                None                  Metal Sheet 7   
347                    LB                  CM         Mountain-100 Black, 38   
503                    LB                  CM    LL Touring Frame - Blue, 58   

                        SpanishProductName  \
586  Montaña: 400 para mujer, plateada, 38   
296                                          
162                                          
347                Montaña: 100, negra, 38   
503           Cuadro de paseo GB: azul, 58   

                            FrenchProductName  StandardCost  \
586                     VTT 400 -W argent, 38        419.78   
296                                                  653.70   
162                                                     NaN   
347                          VTT 100 noir, 38       1898.09   
503  Cadre de vélo de randonnée LL - bleu, 58        199.85   

     FinishedGoodsFlag  ...  \
586                  1  ...   
296                  1  ...   
162                  0  ...   
347                  1  ...   
503                  1  ...   

                                    ChineseDescription  \
586                  此自行车具有优越的性价比。它灵敏且易于操控，越野骑乘也可轻松胜任。   
296  每个车架都是在我们位于 Bothell 的工厂中手工制作，具有高质量山地车架所必需的最佳直径...   
162                                               None   
347  高档竞赛山地车。性能得到进一步增强，包括创新的 HL 车架、极其平稳的前悬架以及适用于所有地...   
503        轻型一体式铝制车架，可以更直立的骑乘姿势进行环城游。我们开创性的设计提供了最佳舒适度。   

                                     ArabicDescription  \
586  توفر هذه الدراجة مستوى عاليًا من الأداء في حدو...   
296  وقد تم تصنيع كل هيكل من تلك الهياكل يدويًا في ...   
162                                               None   
347  دراجة سباقات مخصصة للقيادة في الجبال من أعلى ط...   
503  يوفر هيكل الألومنيوم سميك الأطراف خفيف الوزن و...   

                                     HebrewDescription  \
586  אופניים אלה מספקים רמת ביצועים גבוהה במחיר סבי...   
296  כל מסגרת מיוצרת ידנית במפעל שלנו ב- Bothell, ל...   
162                                               None   
347  אופני הרים תחרותיים מהשורה הראשונה. אפשרויות ל...   
503  מסגרת קלת משקל מאלומיניום חרוץ מספקת תנוחת רכי...   

                                       ThaiDescription  \
586  จักรยานรุ่นนี้มีประสิทธิภาพสูง ในราคาประหยัด  ...   
296  เฟรมแต่ละชิ้นผ่านการสร้างด้วยมือในโรงงาน Bothe...   
162                                               None   
347  สุดยอดจักรยานภูเขาสำหรับการแข่งขัน สมบูรณ์แบบด...   
503  เฟรมอลูมิเนียมน้ำหนักเบา ทำให้ท่วงท่าในการขับข...   

                                     GermanDescription  \
586  Dieses Fahrrad bietet ein erstklassiges Preis-...   
296  "Jeder Rahmen wird in unserem Werk in Braunsfe...   
162                                               None   
347  "Top-Mountainbike für Profis. Mit innovativem ...   
503  Leichter Aluminiumrahmen für aufrechte Sitzhal...   

                                   JapaneseDescription  \
586  この自転車は予算内で高性能を提供します。応答性と操作性に優れ、オフロードでもゆったりした気分...   
296  各フレームは、Bothell 工場で最高級のマウンテン バイクのフレームに必要な最適径および...   
162                                               None   
347  最高級の競技用マウンテン バイクです。性能強化オプションとしては、革新的な HL フレーム、...   
503  軽量接合アルミニウム フレームにより、街乗りに適したアップライトなライディング ポジションを...   

                                    TurkishDescription  StartDate    EndDate  \
586  "Bu bisiklet hem hesaplıdır, hem de yüksek sev... 2013-07-01        NaT   
296  "Her bir kadro Bothell tesisimizde, birinci sı...

### Quali sono i colori disponibili?

In [15]:
df['Color'].unique()

array(['NA', 'Black', 'Silver', 'Red', 'White', 'Blue', 'Multi', 'Yellow',
       'Grey', 'Silver/Black'], dtype=object)

### In media quanto pesano i prodotti?

In [20]:
df['Weight'].mean()

56.1932269503546

### Quanto pesa il più leggero?

In [62]:
df['Weight'].min()

2.12

### Quanto pesa il più pesante?

In [65]:
df['Weight'].max()

1050.0

### Quanti prodotti pesano più di 100 Kg?

In [114]:
kg100=df.loc[:,'Weight']>100           #scrivo la condizionedi estrazione e aggiungo variabile
peso_massimo=df.loc[kg100, 'Weight']   #creo variabile per selezionare sia la condizione che la colonna di interesse
peso_massimo.shape                     #conto quanti valori corrispondono a questa condizione

(28,)

### Quanto costano in media i prodotti (colonna DealerPrice)?

In [141]:
df['DealerPrice'].mean()

448.5970597468353

### Se prendiamo un quarto di tutti i prodotti, in modo che siano i più costosi, quale sarà il loro range di prezzo?

In [121]:
df.describe()   # utilizzo questo metodo e vado a vedere direttamente nella riga 25%: 41.994000 è il valore

ProductKey  ProductSubcategoryKey  StandardCost  FinishedGoodsFlag  \
count   606.00000             397.000000    397.000000         606.000000   
mean    303.50000              12.128463    436.823073           0.655116   
min       1.00000               1.000000      0.860000           0.000000   
25%     152.25000               2.000000     37.120000           0.000000   
50%     303.50000              13.000000    204.630000           1.000000   
75%     454.75000              16.000000    660.910000           1.000000   
max     606.00000              37.000000   2171.290000           1.000000   
std     175.08141               9.423328    497.343079           0.475723   

       SafetyStockLevel  ReorderPoint    ListPrice       Weight  \
count        606.000000    606.000000   395.000000   282.000000   
mean         495.201320    371.400990   747.661544    56.193227   
min            4.000000      3.000000     2.290000     2.120000   
25%          100.000000     75.000000    69.990000     2.680000   
50%          500.000000    375.000000   364.090000    15.420000   
75%         1000.000000    750.000000  1204.320000    24.130000   
max         1000.000000    750.000000  3578.270000  1050.000000   
std          364.007114    273.005336   838.708359   158.000652   

       DaysToManufacture  DealerPrice                      StartDate  \
count         606.000000   395.000000                            606   
mean            1.201320   448.597060  2009-07-26 03:12:28.514851328   
min             0.000000     1.374000            2003-07-01 00:00:00   
25%             0.000000    41.994000            2003-07-01 00:00:00   
50%             1.000000   218.454000            2012-07-01 00:00:00   
75%             1.000000   722.594900            2013-07-01 00:00:00   
max             4.000000  2146.962000            2013-07-01 00:00:00   
std             1.508893   503.225071                            NaN   

                   EndDate  
count                  200  
mean   2008-08-17 14:24:00  
min    2007-12-28 00:00:00  
25%    2007-12-28 00:00:00  
50%    2008-12-27 00:00:00  
75%    2008-12-27 00:00:00  
max    2008-12-27 00:00:00  
std                    NaN

### Qual è il prezzo medio per i prodotti di colore blu?

#### Metodo con query SQL

In [57]:
query_blue='SELECT* FROM dimproduct WHERE COLOR ="Blue"' #query con filtro colore blu
df_blue=pd.read_sql(sql=query_blue, con=db_engine)       #query diventa una variabile per Python
prezzo_medio_blue=df_blue['StandardCost'].mean()         #variabile per creare la condizione di media
"%.2f" % prezzo_medio_blue 

'527.57'

#### Metodo con pd

In [148]:
filtro_blue= df.loc[:,'Color']=='Blue'         #creo un filtro con la condizione richiesta
costi_blue=df.loc[filtro_blue, 'DealerPrice']  #creo una variabile in cui applico il filtro alla colonna che mi interessa
costi_blue.shape                               #mi da quanti prodotti rientrano nel filtro colore blue
costi_blue.mean()                              #applico la funzione media

516.0631785714286

### Qual è il prezzo medio per i prodotti di colore rosso o nero? 

#### Metodo con query SQL

In [66]:
query_color='SELECT* FROM dimproduct WHERE COLOR ="Red" OR COLOR = "Black"' 
df_color=pd.read_sql(sql=query_color, con=db_engine)
prezzo_medio_color=df_color['StandardCost'].mean()
"%.2f" % prezzo_medio_color

'537.45'

#### Metodo con pd

In [176]:
filtro_red=df.loc[:, 'Color']=='Red'               #filtro variabile rosso
filtro_black=df.loc[:, 'Color']=='Black'           #filtro variabile nero
filtro_colori= filtro_red | filtro_black           #unisco i due filtri nella condizione OR

costi_colori=df.loc[filtro_colori, 'DealerPrice']  #variabile a cui applico il filtro sulla colonna che mi interessa
costi_colori.mean()                                #applico la funzione media

550.2481873684211

### Qual è il prezzo massimo per i prodotti di taglia 42 e peso oltre i 10 Kg?

#### Metodo con query SQL

In [78]:
query_size_weight='SELECT * FROM dimproduct WHERE SIZE ="42" AND WEIGHT >10' 
df_size_weight=pd.read_sql(sql=query_size_weight, con=db_engine)
prezzo_max=df_size_weight['StandardCost'].max()
"%.2f" % prezzo_max

'1912.15'

#### Metodo con pd

In [186]:
filtro_taglia=df.loc[:, 'Size']=='42'              #filtro variabile1
filtro_peso=df.loc[:, 'Weight']>10                 #filtro variabile2
filtro= filtro_taglia & filtro_peso                #unisco i due filtri nella condizione 

prezzo_max=df.loc[filtro, 'DealerPrice']           #uso .loc[] con filtro e colonna che mi interessa
prezzo_max.max()                                   #applico la funzione media

2039.994

### Qual è il nome inglese e il costo di produzione (StandardCost) di tutti i prodotti di taglia 42, peso oltre i 10 Kg e colore argento?

#### Metodo con query SQL

In [80]:
query_size_weight_color='SELECT * FROM dimproduct WHERE SIZE ="42" AND WEIGHT >10 AND COLOR="SILVER"'
df_size_weight_color=pd.read_sql(sql=query_size_weight_color, con=db_engine)
df_size_weight_color.loc[:, ['EnglishProductName', 'StandardCost']]

EnglishProductName  StandardCost
0    Mountain-100 Silver, 42       1912.15
1    Mountain-200 Silver, 42       1117.86
2    Mountain-200 Silver, 42       1265.62
3  Mountain-400-W Silver, 42        419.78
4    Mountain-500 Silver, 42        308.22

#### Metodo con pd

In [194]:
filtro_taglia=df.loc[:, 'Size']=='42'                  #filtro variabile1
filtro_peso=df.loc[:, 'Weight']>10                     #filtro variabile2
filtro_colore=df.loc[:, 'Color'] =='Silver'            #filtro variabile3

filtro= filtro_taglia & filtro_peso & filtro_colore    #unisco i filtri nella condizione 
df.loc[filtro,['EnglishProductName', 'StandardCost', 'Size', 'Weight', 'Color']]  

EnglishProductName  StandardCost Size  Weight   Color
344    Mountain-100 Silver, 42       1912.15   42   20.77  Silver
353    Mountain-200 Silver, 42       1117.86   42   23.77  Silver
354    Mountain-200 Silver, 42       1265.62   42   23.77  Silver
588  Mountain-400-W Silver, 42        419.78   42   27.13  Silver
591    Mountain-500 Silver, 42        308.22   42   27.77  Silver

### Esaminiamo la differenza tra lo StandardCost e il DealerPrice degli ultimi 20 elementi del dataset: c'è un pattern? C'è qualche elemento che non lo segue?

In [98]:
print(df[['EnglishProductName', 'StandardCost', 'DealerPrice']].tail(20))

            EnglishProductName  StandardCost  DealerPrice
586  Mountain-400-W Silver, 38        419.78      461.694
587  Mountain-400-W Silver, 40        419.78      461.694
588  Mountain-400-W Silver, 42        419.78      461.694
589  Mountain-400-W Silver, 46        419.78      461.694
590    Mountain-500 Silver, 40        308.22      338.994
591    Mountain-500 Silver, 42        308.22      338.994
592    Mountain-500 Silver, 44        308.22      338.994
593    Mountain-500 Silver, 48        308.22      338.994
594    Mountain-500 Silver, 52        308.22      338.994
595     Mountain-500 Black, 40        294.58      323.994
596     Mountain-500 Black, 42        294.58      323.994
597     Mountain-500 Black, 44        294.58      323.994
598     Mountain-500 Black, 48        294.58      323.994
599     Mountain-500 Black, 52        294.58      323.994
600          LL Bottom Bracket         23.97       32.394
601          ML Bottom Bracket         44.95       60.744
602          H

#### Risposta: normalmente StandardCost è sempre minore di DealerPrice tranne nel caso delle righe 603, 604, 605 in cui la tendenza si inverte